File: java/src/main/java/io/cucumber/prettyformatter/PickleDocStringFormatter.java
package io.cucumber.prettyformatter;

import io.cucumber.messages.types.PickleDocString;

import static io.cucumber.prettyformatter.Theme.Element.DOC_STRING;
import static io.cucumber.prettyformatter.Theme.Element.DOC_STRING_CONTENT;
import static io.cucumber.prettyformatter.Theme.Element.DOC_STRING_DELIMITER;
import static io.cucumber.prettyformatter.Theme.Element.DOC_STRING_MEDIA_TYPE;

final class PickleDocStringFormatter {

    private static final String DOC_STRING_DELIMITER_STRING = "\"\"\"";
    private final int indentation;

    private PickleDocStringFormatter(int indentation) {
        this.indentation = indentation;
    }

    static Builder builder() {
        return new Builder();
    }

    void formatTo(PickleDocString pickleDocString, LineBuilder lineBuilder) {
        lineBuilder
                .indent(indentation)
                .begin(DOC_STRING)
                .append(DOC_STRING_DELIMITER, DOC_STRING_DELIMITER_STRING)
                .accept(lb -> pickleDocString.getMediaType().ifPresent(mediaType -> lb.append(DOC_STRING_MEDIA_TYPE, mediaType)))
                .end(DOC_STRING)
                .newLine();

        // Doc strings are normalized to \n by Gherkin.
        String[] lines = pickleDocString.getContent().split("\\n");
        for (String line : lines) {
            lineBuilder.indent(indentation).
                    begin(DOC_STRING)
                    .append(DOC_STRING_CONTENT, line)
                    .end(DOC_STRING)
                    .newLine();
        }
        lineBuilder
                .indent(indentation)
                .begin(DOC_STRING)
                .append(DOC_STRING_DELIMITER, DOC_STRING_DELIMITER_STRING)
                .end(DOC_STRING)
                .newLine();

    }

    static final class Builder {

        private int indentation = 0;

        private Builder() {

        }

        Builder indentation(int indentation) {
            this.indentation = indentation;
            return this;
        }

        PickleDocStringFormatter build() {
            return new PickleDocStringFormatter(indentation);
        }
    }
}

File: cpp/src/cucumber/pretty-formatter/PickleDocStringFormatter.cpp
#include "cucumber/pretty-formatter/PickleDocStringFormatter.hpp"
#include "cucumber/messages/PickleDocString.hpp"
#include "cucumber/pretty-formatter/LineBuilder.hpp"
#include "cucumber/pretty-formatter/Theme.hpp"
#include <cstddef>
#include <memory>
#include <sstream>
#include <string>

namespace cucumber::pretty_formatter
{
    PickleDocStringFormatter::PickleDocStringFormatter(std::size_t indent)
        : indent{ indent }
    {}

    void PickleDocStringFormatter::Format(LineBuilder& lineBuilder, const std::shared_ptr<const messages::PickleDocString>& docString)
    {
        lineBuilder.Indent(indent)
            .Begin(Theme::Element::docString)
            .Append(Theme::Element::docStringDelimiter, delimiter)
            .Accept(
                [&docString](auto& lineBuilder)
                {
                    if (docString->mediaType.has_value())
                    {
                        lineBuilder.Append(Theme::Element::docStringMediaType, docString->mediaType.value());
                    }
                })
            .End(Theme::Element::docString)
            .NewLine();

        std::istringstream stream{ docString->content };
        for (std::string line; std::getline(stream, line);)
        {
            lineBuilder.Indent(indent)
                .Begin(Theme::Element::docString)
                .Append(Theme::Element::docStringContent, line)
                .End(Theme::Element::docString)
                .NewLine();
        }

        lineBuilder.Indent(indent)
            .Begin(Theme::Element::docString)
            .Append(Theme::Element::docStringDelimiter, delimiter)
            .End(Theme::Element::docString)
            .NewLine();
    }
}
